File: internal/systemd/linger.go
package systemd

import (
	"fmt"
	"os"
	"os/exec"
	"strings"
)

// LingerChecker handles user lingering checks
type LingerChecker struct{}

// NewLingerChecker creates a new linger checker
func NewLingerChecker() *LingerChecker {
	return &LingerChecker{}
}

// IsEnabled checks if lingering is enabled for current user
func (lc *LingerChecker) IsEnabled() (bool, error) {
	username := os.Getenv("USER")
	if username == "" {
		return false, fmt.Errorf("USER environment variable not set")
	}

	cmd := exec.Command("loginctl", "show-user", username, "--property=Linger", "--value")
	output, err := cmd.Output()
	if err != nil {
		// User might not be logged in via loginctl (e.g., in container)
		return false, nil
	}

	return strings.TrimSpace(string(output)) == "yes", nil
}

// WarnIfDisabled prints a warning if lingering is disabled
func (lc *LingerChecker) WarnIfDisabled() {
	enabled, err := lc.IsEnabled()
	if err != nil || enabled {
		return
	}

	username := os.Getenv("USER")
	fmt.Printf("\n⚠️  WARNING: User lingering is not enabled for %s\n", username)
	fmt.Printf("   Your services will stop when you log out.\n")
	fmt.Printf("   To enable lingering (services persist after logout):\n")
	fmt.Printf("   sudo loginctl enable-linger %s\n\n", username)
}

File: src/systemd/linger.rs
use std::process::Command;

pub struct LingerChecker;

impl LingerChecker {
    pub fn new() -> Self {
        Self
    }

    pub fn is_enabled(&self) -> crate::error::Result<bool> {
        let username = std::env::var("USER")
            .map_err(|_| crate::error::Error::msg("USER environment variable not set"))?;
        let output = Command::new("loginctl")
            .args(["show-user", &username, "--property=Linger", "--value"])
            .output();
        let Ok(out) = output else {
            return Ok(false);
        };
        Ok(String::from_utf8_lossy(&out.stdout).trim() == "yes")
    }

    pub fn warn_if_disabled(&self) {
        match self.is_enabled() {
            Ok(true) | Err(_) => {}
            Ok(false) => {
                if let Ok(user) = std::env::var("USER") {
                    println!();
                    println!("⚠️  WARNING: User lingering is not enabled for {user}");
                    println!("   Your services will stop when you log out.");
                    println!("   To enable lingering (services persist after logout):");
                    println!("   sudo loginctl enable-linger {user}");
                    println!();
                }
            }
        }
    }
}

impl Default for LingerChecker {
    fn default() -> Self {
        Self::new()
    }
}
